src/particles.h

```diff
@@ -20,19 +20,269 @@ with this program; if not, write to the Free Software Foundation, Inc.,
+			auto p = numericAbsolute(bias) + 1;
+			for (size_t i = 0; i < sizeof(values) / sizeof(values[0]); ++i) {
+				if (bias < 0)
+					values[i] = 1.0f - pow(myrand_float(), p);
+				else
+					values[i] = pow(myrand_float(), p);
+			}
+			return T::pick(values, min, max);
+		}
+	};
+
+	enum class TweenStyle { fwd, rev, pulse, flicker };
+
+	template <typename T>
+	struct TweenedParameter : public Parameter {
+		using ValType = T;
+		using This = TweenedParameter<T>;
+
+		TweenStyle style = TweenStyle::fwd;
+		u16 reps = 1;
+		f32 beginning = 0.0f;
+
+		T start, end;
+
+		TweenedParameter() = default;
+		TweenedParameter(const This& a) = default;
+		TweenedParameter(T _start, T _end)          : start(_start),  end(_end) {};
+		template <typename M> TweenedParameter(M b) : start(b),       end(b) {};
+
+		T blend(float fac) const {
+			// warp time coordinates in accordance w/ settings
+			if (fac > beginning) {
+				// remap for beginning offset
+				auto len = 1 - beginning;
+				fac -= beginning;
+				fac /= len;
+
+				// remap for repetitions
+				fac *= reps;
+				if (fac > 1) // poor man's modulo
+					fac -= (decltype(reps))fac;
+
+				// remap for style
+				switch (style) {
+					case TweenStyle::fwd: /* do nothing */  break;
+					case TweenStyle::rev: fac = 1.0f - fac; break;
+					case TweenStyle::pulse:
+					case TweenStyle::flicker: {
+						if (fac > 0.5f) {
+							fac = 1.f - (fac*2.f - 1.f);
+						} else {
+							fac = fac * 2;
+						}
+						if (style == TweenStyle::flicker) {
+							fac *= myrand_range(0.7f, 1.0f);
+						}
+					}
+				}
+				if (fac>1.f)
+					fac = 1.f;
+				else if (fac<0.f)
+					fac = 0.f;
+			} else {
+				fac = (style == TweenStyle::rev) ? 1.f : 0.f;
+			}
+
+			return start.interpolate(fac, end);
+		}
+
+		void serialize(std::ostream &os) const override {
+			writeU8(os, (u8)style);
+			writeU16(os, reps);
+			writeF32(os, beginning);
+			start.serialize(os);
+			end.serialize(os);
+		};
+		void deSerialize(std::istream &is) override {
+			u8 st = readU8(is);
+			style = (TweenStyle)st;
+			reps = readU16(is);
+			beginning = readF32(is);
+			start.deSerialize(is);
+			end.deSerialize(is);
+		};
+	};
+
+	// these are consistently-named convenience aliases to make code more readable without `using ParticleParamTypes` declarations
+	using v3fRange = RangedParameter<v3fParameter>;
+	using f32Range = RangedParameter<f32Parameter>;
+
+	// these aliases bind tweened parameter types to the functions used to blend them
+	using v2fTween      = TweenedParameter<v2fParameter>;
+	using v3fTween      = TweenedParameter<v3fParameter>;
+	using f32Tween      = TweenedParameter<f32Parameter>;
+	using v3fRangeTween = TweenedParameter<v3fRange>;
+	using f32RangeTween = TweenedParameter<f32Range>;
+}
+
+struct ParticleTexture {
+	bool animated = false;
+	TileAnimationParams animation;
+	ParticleParamTypes::f32Tween alpha = (f32)1;
+	ParticleParamTypes::v2fTween scale; //= (v2f){1.f,1.f};
```

Comment:
FYI I think this would have worked:
`ParticleParamTypes::v2fTween scale = ParticleParamTypes::v2fTween(v2f(1.f), v2f(1.f));`
or even this perhaps
`ParticleParamTypes::v2fTween scale(1.f, 1.f);`

Reply:
i believe i tried both of those and got errors

Reply:
I don't think brace syntax is a compiler extension

Reply:
brace syntax is not, no. what is considered a compile-time constant expression however is largely up to compilers to define (and we don't have constexpr/consteval in c++11). my compiler and several of the compilers in your test suite accepted the brace syntax as a constant expression, a bunch of others (including the compiler used for android) did not. this was the required fix to enable the code to compile on those compilers.